guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+
+✅
+
+``` typescript
+type WithControllerOptions = Partial<TokensControllerOptions>;
+
+type WithControllerCallback = (setup: {
+  controller: TokensController;
+  onNetworkDidChangeListener: (networkState: NetworkState) => void;
+}) => void | Promise<void>;
+
+type WithControllerArgs =
+  | [WithControllerOptions, WithControllerCallback]
+  | [WithControllerCallback];
+
+async function withController(...args: WithControllerArgs) {
+  const [{
+    chainId = '0x1',
+    onNetworkDidChange = (listener) => {
+      onNetworkDidChangeListener = listener;
+    },
+  }, fn] = args.length === 1 ? [{}, args[0]] : args;
+  const onNetworkDidChangeListener: (networkState: NetworkState) => void;
+
+  const controller = new TokensController({ chainId, onNetworkDidChange });
+  assert(onNetworkDidChangeListener, "onNetworkDidChangeListener was not set")
+
+  try {
+    await fn({ controller, onNetworkDidChangeListener });
+  } catch {
+    controller.destroy();
+  }
+}
+
+describe('TokensController', () => {
+  describe('addToken', () => {
+    it('registers the given token under the default network, assuming it has not changed yet', async () => {
+      await withController({ chainId: '0x1' }, ({ controller }) => {
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x1': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      await withController(({ controller, onNetworkDidChangeListener }) => {
+        onNetworkDidChangeListener({ chainId: '0x42' });
+
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x42': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+  });
+});
+```
+
+The only case in which `beforeEach` or `afterEach` is acceptable is when using the Jest API, such as when setting fake timers:
+
+``` typescript
+describe("some-module", () => {
+  beforeEach(() => {
+    jest.useFakeTimers();
+  });
+
+  afterEach(() => {
+    jest.useRealTimers();
+  });
+
+  // ... tests go here ...
+});
+```
+
+## Keep critical data in the test
+
+Each test involves data that makes that test unique: the test may use special data to construct a scenario, render specific values to its environment, or call out to some other part of the system.
+
+Keeping this data in the test instead of spread out across the file or project makes that test easier to follow from beginning to end.
+
+🚫
+
+``` typescript
+const PASSWORD = 'abc123';
+
+// ... many lines later ...
+
+describe('KeyringController', () => {
+  describe('submitPassword', () => {
+    it('sets the password on the controller', async () => {
+      const keyringController = await initializeKeyringController({
```

Comment:
Hmm. This example doesn't look objectionable to me. The contents of the password aren't really necessary to read this test. As long as you assume it's a non-emtpy string, this is perfectly readable.

Did you intend to omit the password from the initialization options? That would be a much better example. Then we could add a comment explaining that it's not clear to the reader that the password was correct

Reply:
Yes, I think I did mean to do that. I can update this.

Reply:
I also think that this example doesn't illustrate the problem that I'm referring to in the rationale. I actually think there are a couple reasons why it's good to keep critical data in the test. One is that as a reader you don't have to look so far to piece together the "story" the test is telling. The other (which I think you might be referring to here) is that it hides what's special about the data itself. That could be solved by better naming, though. Maybe there's another reason, though. I'll have to study real world code to refine this some more.

Reply:
Yeah exactly. The password variable was completely sufficient for explaining both the "story" of the test and the conditions it was running under, because the test would apply to any arbitrary string as the password. If I had reason to double-check the value of "password" I may have been annoyed at having to scroll all the way up and back down again, but there was no need to look.

Reply:
I went through existing code in `core` and used that to update this example to something I feel like would be 1) more realistic and 2) genuinely annoying. Let me know what you think: [09116f7](https://github.com/MetaMask/contributor-docs/pull/38/commits/09116f79b63c1d0c057965ed451cfc45714edf8b)

Reply:
Oh my goodness, great example, that test is painful